### PR TITLE
Feature/dev onboarding

### PR DESCRIPTION
**Pull request by jkim01**
Added support for wildcard certs, multiple vhosts support, removed version - box version added anisble/group_vars/all, removed all hardcoded settings

### CODE REVIEW

File: lamp/tasks/create_certificate.yml
Comment (likeuntomurphy) [inline lamp/tasks/create_certificate.yml]: I think a better approach here would be to pull out the DNS assignments into their own task that loops through the hosts var and grabs the hostnames since we have them all. The 1-based loop index required for the `DNS.{n}` value should be available as `loop.index`.

File: lamp/tasks/install_php.yml
Comment (likeuntomurphy) [inline lamp/tasks/install_php.yml]: See first comment in the dev-standard review.

File: lamp/tasks/create_certificate.yml
Comment (likeuntomurphy) [inline lamp/tasks/create_certificate.yml]: The use of CN to identify any hostname value has been deprecated, and at least Chrome and Firefox no longer even look there for a hostname — see https://groups.google.com/a/chromium.org/forum/#!msg/security-dev/IGT2fLJrAeo/csf_1Rh1AwAJ. That was the general motivation to just use a generic string value here, as Let’s Encrypt does with their certs (using “Let's Encrypt Authority X3” as the CN value). I think we were only using the hostname as the key and cert name out of habit. But with the possibility of multiple hostnames, we can use something generic, hard-coded and as simple as ssl.crt. Then we can get rid of the `ssl_cert_name` var completely.

**Review by likeuntomurphy — CHANGES_REQUESTED** [review]
As a general comment, I’d also like to see a way to have the multiple-host feature be opt-in. It clearly makes sense for something like screening-ee that required screening-admin, but perhaps less so to keep  projects grouped simply because they’re for the same client.

**Comment by jkim01** [comment]
Ah, good stuff. I just pushed the updates to install_php and create_certificate, but haven't isolated the DNS assignments into their own tasks yet. Will get to that soon as well as items in the dev-standards repo.

What are your thoughts on moving all of the required php and mysql modules from ansible/group_vars/all into their own tasks? I ask because I'll be working with Mike on Python projects and while going through this whole process I was also trying to plan out creating Python/Django/Flask, etc environments using the current standards in place. Now that I know the trusty branch exists, I was hoping to get ansible/group_vars/all to only contain general project information such as local path(s), guest web root location(s), etc.

**Comment by jkim01** [comment]
Alright, Jonathan. Got some updates out on both repos in the feature/dev-onboarding branch. Let me know what you think and your thoughts on putting the php_modules list from group_vars/all into the install_php playbook.

**Comment by likeuntomurphy** [comment]
I think we’re more or less on the same page regarding the `provision.yml` file and `vars/all` because they are meant to be project-speficic — so any PHP modules listed there should be required by the current project but not necessarily by every project. Because of this, I’m reluctant to move them into `install_php.yml` and hence load them in every project. Of course, one could argue that there’s no harm in installing a handful of modules in every project and expanding the list as projects need new modules — but I think maintaining a list in the project’s vars file also helps developers get a quick glimpse of a project’s “special requirements.”

And while the default `provision.yml` appears to be biased toward a LAMP stack, you can do whatever you want with it in conjunction with adding new roles/tasks to this repository. For any project, you can remove the Yaml entry for `php_modules` entirely, and there will be no error — even if you run the LAMP role, since `php_modules` defaults to an empty array. So naturally you can just remove the call for the LAMP role and add additional variables like `python_packages` to `provision.yml` at will, so long as their absence doesn’t create errors.

With all that said, it seems like it boils down simply identifying which variables are required and which are optional in `vars/all`, which is more of a documentation issue.

**Comment by jkim01** [comment]
Yeah, that makes sense leaving the php_modules in vars/all. I also agree that it's a documentation issue with projects.

Cool, I've been working on the Django/Python stuff on the side so once you're OK with the changes in place to support multiple hosts, etc., I'll merge this branch into master after fixing the conflict and then create another pull request for the Django/Python stuff (think this is going to be a work in progress as Mike and I start working on existing/new projects).

**Comment by jkim01** [comment]
Hey Jonathan. Just curious to know if there are any additional changes you want made to this branch so that I can merge it along with the changes here (https://github.com/fostermadeco/development-standard/tree/feature/dev-onboarding) into their respective master branches.

**Comment by alexmglover** [comment]
@jkim01 Since Jonathan is out of town this week I thought I would respond. Due to the size of this change, he'll likely need to do another thorough code review before merging this into master because of how significant the changes are.

File: lamp/tasks/create_certificate.yml
Comment (likeuntomurphy) [inline lamp/tasks/create_certificate.yml]: The vhosts template also needs the name updated.

**Comment by jkim01** [comment]
Weird, thought all that had been reverted. Just pushed out the fixes to the LAMP provisioner. If you could let me know before you merge this into master, I'd appreciate it. Would like to run one final test setting up the screening admin and screening-ee sites.

**Comment by likeuntomurphy** [comment]
I think this will be good to go once`development/standard` is ready.

**Comment by jkim01** [comment]
Ok, Jonathan. There seems to be a conflict with setting config.vm.hostname while using config.hostsupdater.aliases in the Vagrant file; it'll create duplicate entries in one's local /etc/hosts. To get around this, I'm telling Ansible to create the hostname by way of a new task, set_hostname.yml, which has been added to system/tasks/main.yml. If you're cool with this, let me know and I'll run one final test setting up a single host box and a multi-host box before merging this and the development-standards into master.

**Comment by likeuntomurphy** [comment]
What if we change the head of the configure block in Vagrantfile to,

```ruby
  config.vm.hostname = vars["hosts"][0]["hostname"]

  if vars["hosts"].count > 1
    config.hostsupdater.aliases = vars["hosts"][1..-1].map{|host| host["hostname"]}

    config.trigger.before :up do
      run "ansible-playbook ansible/clone_repositories.yml -i localhost," # comma is necessary
    end
  end
```

It will still create two lines in `/etc/hosts`, but the hostname will only be present in one line.

**Comment by jkim01** [comment]
Cool, this works. Just tested on this on a single and multi-host machine. Will remove the additions I made to this branch and update the Vagrantfile in dev-standards

**Comment by jkim01** [comment]
While going through this for screening-admin-l5 and screening-ee there were instances where I had to update some roles, which I don't believe the developer should have to, so they've been updated. 

- mysql-server and mysql-client have been removed from install_mysql.yml. The version of Ubuntu (trusty) which Screening is using supports multiple versions of mysql and by default it selects 5.5 which doesn't support full text indexes. Luckily, 5.6 is available for installation. These variables are now set in group_vars/all.

- zip has been removed from install_php.yml and placed into group_vars/all; there is no php5-zip package. 

Unfortunately, for Screening, install_composer.yml has to be edited. Perhaps since this is only local development it can be added to the task, permanently. Even after installing the necessary python packages, `get composer installer hash` fails. The only way around it was to add `validate_certs: no`.

Here is the official/unofficial client documentation for Screening:

https://s4mh.beanstalkapp.com/screening-admin-l5/browse/git/readme.md?ref=b-feature%2Fdev-onboarding#ansible-group_vars-updates-header

**Comment by likeuntomurphy** [comment]
@jkim01 If you're going to be using Docker exclusively for Python projects, can we strip out the Python changes and keep just the multi-host functionality?

**Comment by jkim01** [comment]
@likeuntomurphy That should be fine. If a future client doesn't use Docker, I have a project using the Ansible provisioner @fmmike and I can use.

**Pull request closed without merging**
